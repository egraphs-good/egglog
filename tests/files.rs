use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
