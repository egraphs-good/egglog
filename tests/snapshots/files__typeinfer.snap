---
source: tests/files.rs
expression: serialized
---
nodes:
  TInt-0:
    op: TInt
    children: []
    eclass: "0"
    cost: 1
  TBool-0:
    op: TBool
    children: []
    eclass: "1"
    cost: 1
  TUnit-0:
    op: TUnit
    children: []
    eclass: "2"
    cost: 1
  Unit-0:
    op: ()
    children: []
    eclass: Unit-0
    cost: 0
  base-types-0:
    op: base-types
    children:
      - TInt-0
    eclass: Unit-0
    cost: 1
  base-types-5871781006564002453:
    op: base-types
    children:
      - TBool-0
    eclass: Unit-0
    cost: 1
  base-types-11743562013128004906:
    op: base-types
    children:
      - TUnit-0
    eclass: Unit-0
    cost: 1
root_eclasses: []
class_data:
  "0":
    type: Type
  "1":
    type: Type
  "2":
    type: Type
  Unit-0:
    type: Unit

